File: Exercise_3.cpp
/* 
    Time Complexity                              :  O(N)
    Space Complexity                             :  Recursive Approach - O(N) when the tree is skewed
                                                    Backtracking - O(H + H + H + H) 
                                                    Memory consumed due to the recursion stack and 
                                                    memory due to the pathP, pathQ and path.
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return recursive(root, p, q);
        // vector<TreeNode*> pathQ;
        // vector<TreeNode*> pathP;
        // backtracking(root, p, q, vector<TreeNode*>(), pathP, pathQ);
        // for(int i=0;i<pathP.size();i++) {
        //     if(pathP[i] != pathQ[i]) {
        //         return pathP[i-1];
        //     }
        // }
        // return NULL;
    }
    
    TreeNode* recursive(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == NULL or root == p or root == q) return root;
        TreeNode *ra = lowestCommonAncestor(root->right,p,q), *la = lowestCommonAncestor(root->left, p, q);
        if(la and ra) return root;
        return la?la:ra;
    }
    
    // memory limit exceeded
    void backtracking(TreeNode* root, TreeNode* p, TreeNode* q, vector<TreeNode*> path, vector<TreeNode*>& pathP, vector<TreeNode*>& pathQ) {
        if(!root) return;
        
        path.push_back(root);
        if(root == p) {
            pathP = path;
            pathP.push_back(root);
        }
        
        if(root == q) {
            pathQ = path;
            pathQ.push_back(root);
        }
        
        backtracking(root->left, p, q, path, pathP, pathQ);
        backtracking(root->right, p, q, path, pathP, pathQ);
        
        path.pop_back();
    }
};

File: Exercise_3.py
#Time Complexity : O(N), N being the number of nodes in the tree. 
#Space Complexity : O(h), h being the height of the tree.  
#Did this code successfully run on Leetcode : Yes.
#Any problem you faced while coding this : Had to look at solution to find the strategy. 
'''Find the path to the two nodes using backtracking. Then, recurse through the two 
paths until the values are not the same. Return the index-1. 
'''
class Solution:
    def __init__(self):
        self.pathp=[]
        self.pathq=[]

    # find the two paths using backtracking. 
    def helper(self, root, p, q, path):
        path.append(root)
        if root==p:
            self.pathp = copy.deepcopy(path)
        if root==q:
            self.pathq = copy.deepcopy(path)
        if root.left:
            self.helper(root.left, p, q, path)
        if root.right:
            self.helper(root.right, p, q, path)
        path.pop(len(path)-1)

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root==None:
            return root
        
        index = 0
        self.helper(root, p, q, [])
        
        #loop through the paths until the values do not equal. 
        length = max(len(self.pathp), len(self.pathq))
        for i in range(length):
            if len(self.pathq)-1<i:
                index = i
                break
            if len(self.pathp)-1<i:
                index = i
                break
            p_node = self.pathp[i]
            q_node = self.pathq[i]
            if p_node.val!=q_node.val:
                index = i
                break

        return self.pathp[index-1]